apis/core/v1alpha2/featuregate_webhook.go

```diff
@@ -0,0 +1,270 @@
+
+var _ webhook.Validator = &FeatureGate{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *FeatureGate) ValidateCreate() error {
+	featuregatelog.Info("validate create", "name", r.Name)
+
+	ctx := context.Background()
+	var allErrors field.ErrorList
+
+	c, err := r.getClient()
+	if err != nil {
+		return apierrors.NewInternalError(err)
+	}
+
+	allErrors = append(allErrors, r.validateFeatureExistence(ctx, c)...)
+	allErrors = append(allErrors, r.validateFeatureConflicts(ctx, c)...)
+	allErrors = append(allErrors, r.validateFeatureForStabilityPolicyViolation(ctx, c)...)
+	if len(allErrors) == 0 {
+		return nil
+	}
+
+	return apierrors.NewInvalid(GroupVersion.WithKind("FeatureGate").GroupKind(), r.Name, allErrors)
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (r *FeatureGate) ValidateUpdate(old runtime.Object) error {
+	featuregatelog.Info("validate update", "name", r.Name)
+
+	oldObj, ok := old.(*FeatureGate)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected FeatureGate object, but got object of type %T", old))
+	}
+	if oldObj == nil {
+		return nil
+	}
+
+	c, err := r.getClient()
+	if err != nil {
+		return apierrors.NewInternalError(err)
+	}
+
+	ctx := context.Background()
+	var allErrors field.ErrorList
+
+	allErrors = append(allErrors, r.validateFeatureExistence(ctx, c)...)
+	allErrors = append(allErrors, r.validateFeatureConflicts(ctx, c)...)
+	allErrors = append(allErrors, r.validateStabilityPolicyOverride(oldObj)...)
+	allErrors = append(allErrors, r.validateFeatureForStabilityPolicyViolation(ctx, c)...)
+
+	if len(allErrors) == 0 {
+		return nil
+	}
+	return apierrors.NewInvalid(GroupVersion.WithKind("FeatureGate").GroupKind(), r.Name, allErrors)
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (r *FeatureGate) ValidateDelete() error {
+	featuregatelog.Info("validate delete", "name", r.Name)
+	return nil
+}
+
+// validateFeatureForStabilityPolicyViolation validates the feature for any stability policy violation
+func (r *FeatureGate) validateFeatureForStabilityPolicyViolation(ctx context.Context, c client.Client) field.ErrorList {
+	var allErrors field.ErrorList
+
+	features := &FeatureList{}
+	if err := c.List(ctx, features); err != nil {
+		allErrors = append(allErrors, field.InternalError(field.NewPath("spec").Child("features"), err))
+		return allErrors
+	}
+
+	invalidFeatures := sets.String{}
+	for _, featureRef := range r.Spec.Features {
+		stabilityLevel, found := getFeatureStabilityLevel(features, featureRef.Name)
+		if !found {
+			// Feature doesn't exist and is validated in validateFeatureExistence method
+			continue
+		}
+		policy := GetPolicyForStabilityLevel(stabilityLevel)
+		// immutability validation
```

Comment:
Optional: I don't think this comment is very helpful. I can tell from line 146 this is some sort of validation on `policy.Immutable`. What would be helpful is a comment that says something like, "Verify that a feature that is immutable due to the stability policy has sufficient overrides to be (de)activated."